<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Compare row ordering methods • EnrichedHeatmap</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Compare row ordering methods">
<meta property="og:description" content="EnrichedHeatmap">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">EnrichedHeatmap</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.29.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Vignettes</li>
    <li>
      <a href="../articles/EnrichedHeatmap_intro.html">Make Enriched Heatmaps</a>
    </li>
    <li>
      <a href="../articles/visualize_categorical_signals.html">Visualize Categorical Signals</a>
    </li>
    <li>
      <a href="../articles/row_ordering.html">Compare row ordering methods</a>
    </li>
    <li>
      <a href="../articles/roadmap.html">Visualize Comprehensive Associations in Roadmap dataset</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">More applications</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/EnrichedHeatmap/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Compare row ordering methods</h1>
                        <h4 data-toc-skip class="author">Zuguang Gu ( <a href="mailto:z.gu@dkfz.de" class="email">z.gu@dkfz.de</a> )</h4>
            
            <h4 data-toc-skip class="date">2024-02-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/EnrichedHeatmap/blob/HEAD/vignettes/row_odering.Rmd" class="external-link"><code>vignettes/row_odering.Rmd</code></a></small>
      <div class="hidden name"><code>row_odering.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/EnrichedHeatmap" class="external-link">EnrichedHeatmap</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"neg_cr.RData"</span>, package <span class="op">=</span> <span class="st">"EnrichedHeatmap"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">all_genes</span> <span class="op">=</span> <span class="va">all_genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">neg_cr</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">all_tss</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">promoters</a></span><span class="op">(</span><span class="va">all_genes</span>, upstream <span class="op">=</span> <span class="fl">0</span>, downstream <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">mat_neg_cr</span> <span class="op">=</span> <span class="fu"><a href="../reference/normalizeToMatrix.html">normalizeToMatrix</a></span><span class="op">(</span><span class="va">neg_cr</span>, <span class="va">all_tss</span>, mapping_column <span class="op">=</span> <span class="st">"gene"</span>, w <span class="op">=</span> <span class="fl">50</span>, mean_mode <span class="op">=</span> <span class="st">"w0"</span><span class="op">)</span></span></code></pre></div>
<p>The object <code>mat_neg_cr</code> is a normalized matrix for regions showing significant negative correlation between methylation and gene expression. The negative correlated regions (negCRs) are normalized to upstream 5kb and downstream 5kb of gene TSS with 50bp window by <code><a href="../reference/normalizeToMatrix.html">normalizeToMatrix()</a></code> function. The value in the matrix is how much a window is covered by negCRs (values between 0 and 1).</p>
<p>In the normalized matrix, each row corresponds to one gene and each column corresponds to a window either on upstream of TSS or downstream of TSS. For the example of <code>mat_neg_cr</code> matrix, the first half columns correspond to the upstream of TSS and the last half columns correspond to downstream of TSS. Here we compare following three different methods to order rows (which correspond to genes) in the normalized matrix.</p>
<ol style="list-style-type: decimal">
<li>
<p>Rows are ordered by the enriched scores. For each row in the matrix, denote values in a certain row as <span class="math inline">\(x\)</span>, indices 1, …, <span class="math inline">\(n_1\)</span> are for upstream windows, indices <span class="math inline">\(n_1+1\)</span>, …, <span class="math inline">\(n\)</span> are for downstream windows and <span class="math inline">\(n_2 = n - n_1\)</span>, the enriched score is calculated as the sum of <span class="math inline">\(x\)</span> weighted by distance to TSS (higher weight if the window is close to TSS).</p>
<p><span class="math display">\[ \sum_{i=1}^{n_1}{x_i \cdot i/n_1} + \sum_{i=n_1+1}^n{x_i \cdot (n - i + 1)/n_2}\]</span></p>
</li>
<li><p>Rows are ordered by hierarchical clustering with Euclidean distance.</p></li>
<li>
<p>Rows are ordered by hierarchical clustering with closeness distance. For two rows in the normalized matrix, assume <span class="math inline">\(a_1, a_2, ..., a_{n_1}\)</span> are the indices of windows for one gene which overlap with negCRs and <span class="math inline">\(b_1, b_2, ... b_{n_2}\)</span> are the indices for the other gene which overlap with negCRs, the distance which is based on closeness of the overlapped windows in the two genes is defined as:</p>
<p><span class="math display">\[ d_{closeness} = \frac{\sum_{i=1}^{n_1} \sum_{j=1}^{n_2} {|a_i - b_j|} }{n_1 \cdot n_2}\]</span></p>
<p>So the closeness distance is basically the average distance of all pairs of negCR windows in the two genes.</p>
</li>
</ol>
<p>Euclidean distance between rows keeps unchanged when the matrix columns are permutated, while for closeness distance, the column order is also taken into account, which might be more proper for clustering normalized matrices because the columns correspond to relative distance to the target regions.</p>
<p>Following three plots show heatmaps under different row ordering methods.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/EnrichedHeatmap.html">EnrichedHeatmap</a></span><span class="op">(</span><span class="va">mat_neg_cr</span>, name <span class="op">=</span> <span class="st">"neg_cr"</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"darkgreen"</span><span class="op">)</span>, </span>
<span>    top_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html" class="external-link">HeatmapAnnotation</a></span><span class="op">(</span>enrich <span class="op">=</span> <span class="fu"><a href="../reference/anno_enriched.html">anno_enriched</a></span><span class="op">(</span>gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="st">"darkgreen"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    row_title <span class="op">=</span> <span class="st">"by default enriched scores"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/EnrichedHeatmap.html">EnrichedHeatmap</a></span><span class="op">(</span><span class="va">mat_neg_cr</span>, name <span class="op">=</span> <span class="st">"neg_cr"</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"darkgreen"</span><span class="op">)</span>,</span>
<span>    top_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html" class="external-link">HeatmapAnnotation</a></span><span class="op">(</span>enrich <span class="op">=</span> <span class="fu"><a href="../reference/anno_enriched.html">anno_enriched</a></span><span class="op">(</span>gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="st">"darkgreen"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    row_title <span class="op">=</span> <span class="st">"by hierarchcal clustering + Euclidean distance\ndendrogram reordered by enriched scores"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/EnrichedHeatmap.html">EnrichedHeatmap</a></span><span class="op">(</span><span class="va">mat_neg_cr</span>, name <span class="op">=</span> <span class="st">"neg_cr"</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"darkgreen"</span><span class="op">)</span>,</span>
<span>    top_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html" class="external-link">HeatmapAnnotation</a></span><span class="op">(</span>enrich <span class="op">=</span> <span class="fu"><a href="../reference/anno_enriched.html">anno_enriched</a></span><span class="op">(</span>gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="st">"darkgreen"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>, clustering_distance_rows <span class="op">=</span> <span class="va">dist_by_closeness</span>,</span>
<span>    row_title <span class="op">=</span> <span class="st">"by hierarchcal clustering + closeness distance\ndendrogram reordered by enriched scores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="row_odering_files/figure-html/unnamed-chunk-4-1.png" width="768"></p>
<p>Generally, when the top annotation which summarises mean enrichment across genes is also added to the heatmap, ordering genes by enriched scores is not recommended because it provides redundant information as the top enriched annotation, and on the other hand, it fails to reveal spatial clusters as other two methods. Hierarchal clustering with Euclidean distance is good at clustering enrichment patterns, but since it does not take column order into account, thus, it still can be possible that two spatially close clusters are far separated in the heatmap. By using closeness distance, it clearly sorts and clusters the enrichment patterns.</p>
<p>The row order, clustering method, distance method can all be self-adjusted by <code>row_order</code>, <code>cluster_rows</code>, <code>clustering_method_rows</code>, <code>clustering_distance_rows</code> arguments in <code>EnrichedHeamtap()</code> function. For how to properly set values for these arguments, users can go to the help page of <code><a href="../reference/EnrichedHeatmap.html">EnrichedHeatmap()</a></code> or <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap()</a></code> function.</p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  <style>nav[data-toggle='toc'] .nav .nav {display: block;}</style>
</body>
</html>
